### PR TITLE
Check if previous/next page is hidden to show previous/next buttons

### LINKED ISSUE #2568: Handling previous button when first item has no ‘page’ and hidden
**Describe the bug**
Sometimes the "previous" button used for paginations shows in the first page, and crashes with the following exception when clicked

```txt
org...rtregister.fhircore.mwcoreDev  E  FATAL EXCEPTION: main
                                                                                                    Process: org.smartregister.fhircore.mwcoreDev, PID: 16413
                                                                                                    java.lang.IllegalStateException: Can't call goToPreviousPage() if no preceding page is enabled
                                                                                                    	at com.google.android.fhir.datacapture.QuestionnaireViewModel.goToPreviousPage$datacapture_release(QuestionnaireViewModel.kt:483)
                                                                                                    	at com.google.android.fhir.datacapture.QuestionnaireFragment.onViewCreated$lambda$1(QuestionnaireFragment.kt:98)
                                                                                                    	at com.google.android.fhir.datacapture.QuestionnaireFragment.$r8$lambda$Uy3n-EtWjyq7ZdNYJNMoLcmE2DA(Unknown Source:0)
                                                                                                    	at com.google.android.fhir.datacapture.QuestionnaireFragment$$ExternalSyntheticLambda0.onClick(Unknown Source:2)
                                                                                                    	at android.view.View.performClick(View.java:7448)
                                                                                                    	at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1218)
                                                                                                    	at android.view.View.performClickInternal(View.java:7425)
                                                                                                    	at android.view.View.access$3600(View.java:810)
                                                                                                    	at android.view.View$PerformClick.run(View.java:28305)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loop(Looper.java:223)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7656)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Component**
SDC library

**Expected behavior**
Previous button shouldn't show up on the first page

**Screenshots**

**Additional context**
It seems to happen when the questionnaire has items (in our case, hidden)  not marked with extension for page, coming before ones with page

```txt
...

  "item": [
    {
      "linkId": "just-hidden",
      "extension": [
        {
          "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-hidden",
          "valueBoolean": true
        }
      ],
      "type": "boolean",
      "initial": [
        {
          "valueBoolean": true
        }
      ]
    },
    {
      "linkId": "PR",
      "type": "group",
      "text": "Client info",
      "extension": [
        {
          "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
          "valueCodeableConcept": {
            "coding": [
              {
                "system": "http://hl7.org/fhir/questionnaire-item-control",
                "code": "page",
                "display": "Page"
              }
            ]
          }
        }
      ],

....
``` 
[example.json](https://github.com/user-attachments/files/15805896/new-patient-registration-paginated.json)

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by LZRS**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2568 

**Description**
Change `getQuestionnairePages` to only return questionnaire pages from QuestionnaireItems that have extension 
```
        {
          "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
          "valueCodeableConcept": {
            "coding": [
              {
                "system": "http://hl7.org/fhir/questionnaire-item-control",
                "code": "page",
                "display": "Page"
              }
            ]
          }
        }
```

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Bug fix

**Screenshots (if applicable)**

**Checklist**

- [ ] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [ ] I have read the [Contributing](https://google.github.io/android-fhir/contrib/) page.
- [ ] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [ ] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [ ] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [ ] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 1]
```diff
@@ -480,9 +481,7 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
       EntryMode.PRIOR_EDIT,
       EntryMode.RANDOM, -> {
         val previousPageIndex =
-          pages!!.indexOfLast {
-            it.index < currentPageIndexFlow.value!! && it.enabled && !it.hidden
-          }
+          pages!!.last { it.index < currentPageIndexFlow.value!! && it.enabled && !it.hidden }.index
```
Comment (MJ1998) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:484]: How are these 2 different ?

**Review by MJ1998 — CHANGES_REQUESTED** [review]
We need to include contiguous set of items not with page extension into another page.

"If there are items at the same level as a 'page' group that are listed before the 'page' group, they will be treated as a separate page"
Reference - https://build.fhir.org/ig/HL7/fhir-extensions//CodeSystem-questionnaire-item-control.html

So if the questionnaire looks like - 
{
  itemWithNoPageExtension1, 
  itemWithPageExtension1, // have nested items
  itemWithPageExtension2,
  itemWithNoPageExtension2,
  itemWithNoPageExtension3,
  itemWithNoPageExtension4,
  itemWithPageExtension3,
  }
  
The pages would look like following:-
{
  page1: {
  	itemWithNoPageExtension1,
  },
  page2: {
  	itemWithPageExtension1
  },
  page3: {
  	itemWithPageExtension2
  },
  page4: {
  	itemWithNoPageExtension2,
  	itemWithNoPageExtension3,
  	itemWithNoPageExtension4
  },
  page5: {
  	itemWithPageExtension3
  }
}

**Comment by FikriMilano** [comment]
@LZRS does this change handle the following use case?
```
if the first page is by default, hidden using an extension, while the user is in the second page, the previous button should be hidden
```

**Comment by LZRS** [comment]
> @LZRS does this change handle the following use case?
> 
> ```
> if the first page is by default, hidden using an extension, while the user is in the second page, the previous button should be hidden
> ```

Yes, so far the changes in this PR handle that but the ongoing discussions were on whether it should be hidden since it's also treated as a 'page' or should we handle by through an exception

**Comment by FikriMilano** [comment]
@LZRS 
I mean, what if the first page has the page extension? but the page item is hidden using an extension.

In the second page, which also has a page extension, the previous button needs to be hidden.

Can this change handle that?

**Comment by LZRS** [comment]
> @LZRS I mean, what if the first page has the page extension? but the page item is hidden using an extension.
> 
> In the second page, which also has a page extension, the previous button needs to be hidden.
> 
> Can this change handle that?

Yeah, it handles that

**Comment by FikriMilano** [comment]
Here's some updates about the issue in dealing with pages, by @MJ1998 
https://chat.fhir.org/#narrow/channel/179255-questionnaire/topic/Rendering.20of.20pages.20when.20page.20extension.20is.20absent.20in.20items/near/481103016

**Comment by FikriMilano** [comment]
@MJ1998 @jingtang10 have you decided whether we should make the implementation more relaxed?

**Comment by LZRS** [comment]
@FikriMilano @MJ1998 Reviewing the current existing implementation,  **Sibling** items to a page extension group are treated as page even though they may not be of type`group`. 
```
{
  page1: {
  	itemWithNoPageExtension1
  },
  page2: {
  	itemWithPageExtension2
  },
  page3: {
  	itemWithPageExtension3
  },
  page4: {
  	itemWithNoPageExtension4
},
page5: {
  	itemWithNoPageExtension5
},
page6: {
  	itemWithNoPageExtension6
  },
  page7: {
  	itemWithPageExtension7
  }
}
```
The [spec](https://build.fhir.org/ig/HL7/fhir-extensions//CodeSystem-questionnaire-item-control.html) has also been updated 

> If a group item has this itemControl value, then all sibling items SHALL also be of type group and also have an itemControl value of 'page', 'header', or 'footer'. Form fillers SHOULD raise an error if a Questionnaire fails to adhere to this requirement.

IMO, it feels the current implementation is already relaxed and wouldn't necessarily crash 

In the PR, I have just updated to check that the previous or next button is not hidden, to show the previous/next button

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 2]
```diff
@@ -1533,6 +1533,102 @@ class QuestionnaireViewModelTest {
...
+                linkId = "page3-1"
+                type = Questionnaire.QuestionnaireItemType.BOOLEAN
+                text = "Question on page 3"
+              },
+            )
+          },
+        )
+        addItem(
+          QuestionnaireItemComponent().apply {
+            linkId = "page4-noExtension"
```
Comment (MJ1998) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:1583]: NIT
```suggestion
            linkId = "page4-noExtension-hidden"
```

**Review by MJ1998 — APPROVED** [review]
LGTM. Approving with one small NIT.

**Pull request merged**